### PR TITLE
ChainlinkL2USDMultiOracle: Chainlink USD Oracle that's safe to use on L2

### PR DESCRIPTION
**Pull request by sblOWPCKCR**
Incorporates https://docs.chain.link/docs/l2-sequencer-flag/

Instead of parametrizing ChainlinkUSDMultiOracle, we create new contract to avoid
increasing the gas cost of ChainlinkUSDMultiOracle when there's no need to check the
sequencer flag

### CODE REVIEW

File: contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol [hunk 1]
```diff
@@ -104,3 +105,57 @@ contract ChainlinkUSDMultiOracle is IOracle, AccessControl, Constants {
...
+    // https://docs.chain.link/docs/l2-sequencer-flag/
+    address internal constant FLAG_ARBITRUM_SEQ_OFFLINE =
+        address(bytes20(bytes32(uint256(keccak256('chainlink.flags.arbitrum-seq-offline')) - 1)));
+
+    constructor(FlagsInterface flags) {
+        require(address(flags) != address(0), 'FlagsInterface has to be set');
+        chainlinkFlags = flags;
+    }
+
+    function assertCanGetPrice() private view {
```
Comment (alcueca) [inline contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol:129]: This looks like a good candidate for a modifier:
```
modifier onlyFresh() {
  // If flag is raised we shouldn't perform any critical operations
  require(
    chainlinkFlags.getFlag(FLAG_ARBITRUM_SEQ_OFFLINE) != isRaised,
    "Chainlink feeds are not being updated"
  );
  _;
}
```

File: contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol [hunk 2]
```diff
@@ -104,3 +105,57 @@ contract ChainlinkUSDMultiOracle is IOracle, AccessControl, Constants {
...
+    function peek(
+        bytes32 baseId,
+        bytes32 quoteId,
+        uint256 amountBase
+    ) external view virtual override returns (uint256 amountQuote, uint256 updateTime) {
+        assertCanGetPrice();
+        if (baseId == quoteId) return (amountBase, block.timestamp);
+
+        (amountQuote, updateTime) = _peekThroughUSD(baseId.b6(), quoteId.b6(), amountBase);
+    }
```
Comment (alcueca) [inline contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol:147]: Simpler code:
```
    /// @dev Convert amountBase base into quote at the latest oracle price.
    function peek(bytes32 baseId, bytes32 quoteId, uint256 amountBase)
      external view virtual override
      onlyFresh
      returns (uint256 amountQuote, uint256 updateTime)
    {
      return super.peek(baseId, quoteId, amountBase);
    }
```

File: contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol [hunk 3]
```diff
@@ -104,3 +105,57 @@ contract ChainlinkUSDMultiOracle is IOracle, AccessControl, Constants {
...
+        bytes32 quoteId,
+        uint256 amountBase
+    ) external virtual override returns (uint256 amountQuote, uint256 updateTime) {
+        assertCanGetPrice();
+
+        if (baseId == quoteId) return (amountBase, block.timestamp);
+
+        (amountQuote, updateTime) = _peekThroughUSD(baseId.b6(), quoteId.b6(), amountBase);
+    }
+}
```
Comment (alcueca) [inline contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol:161]: As before:
```
    function get(bytes32 baseId, bytes32 quoteId, uint256 amountBase)
      external virtual override
      onlyFresh
      returns (uint256 amountQuote, uint256 updateTime)
    {
      return super.get(baseId, quoteId, amountBase);
    }
```

**Review by alcueca — CHANGES_REQUESTED** [review]
Fine in principle, but the code can be simpler. Interesting stuff.

File: contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol (hunk 2 above)
Comment (sblOWPCKCR) [reply contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol:147]: nope, can't do that for `external` functions

File: contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol (hunk 3 above)
Comment (sblOWPCKCR) [reply contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol:161]: nope, can't do that for `external` functions

File: contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol (hunk 1 above)
Comment (sblOWPCKCR) [reply contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol]: I find modifiers easy to miss, so went for an explicit check in the function body
Also, remember Tenderly always pointing you at `auth` modifier as the source of the revert? That's another problem with modifiers

With that said, I'm not super opposed to them. Took your suggestion

File: contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol [hunk 4]
```diff
@@ -104,3 +105,52 @@ contract ChainlinkUSDMultiOracle is IOracle, AccessControl, Constants {
...
 }
+
+/**
+ * @title ChainlinkL2USDMultiOracle: ChainlinkUSDMultiOracle that's safe to use on L2
+ * @notice Chainlink recommends checking the sequencer status on some L2 networks to avoid
+  reading stale data
+
+  https://docs.chain.link/docs/l2-sequencer-flag/
+ */
+contract ChainlinkL2USDMultiOracle is ChainlinkUSDMultiOracle {
```
Comment (alcueca) [inline contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol:116]: Didn't notice this before, could you please split ChainlinkL2USDMultiOracle into its own file?

**Review by alcueca — CHANGES_REQUESTED** [review]
One contract per file, please.

File: contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol (hunk 2 above)
Comment (alcueca) [reply contracts/oracles/chainlink/ChainlinkUSDMultiOracle.sol]: Oh well

**Pull request merged**